exonum/src/api/node/public/explorer.rs

```diff
@@ -246,55 +252,112 @@ impl ExplorerApi {
                 ApiError::NotFound(description)
             })
     }
+
     /// Adds transaction into unconfirmed tx pool, and broadcast transaction to other nodes.
     pub fn add_transaction(
+        name: &str,
+        backend: &mut actix_backend::ApiBuilder,
+        service_api_state: ServiceApiState,
+    ) {
+        let (max_message_len, max_payload_len) = get_message_limits(&service_api_state);
+        let index = move |request: HttpRequest| {
+            let state = request.state().clone();
+            let content_length = request
+                .headers()
+                .get(CONTENT_LENGTH)
+                .unwrap()
+                .to_str()
+                .unwrap_or_default()
+                .to_owned();
+            request
+                .json()
+                .limit(max_payload_len)
+                .map_err(move |e| convert_error(e, max_payload_len, content_length))
+                .from_err()
+                .and_then(move |query: TransactionHex| {
+                    Self::tx_handler(&state, query, max_message_len)
+                        .map(|value| HttpResponse::Ok().json(value))
+                        .map_err(From::from)
+                })
+                .responder()
+        };
+
+        backend.raw_handler(RequestHandler {
+            name: name.to_owned(),
+            method: http::Method::POST,
+            inner: Arc::new(index) as Arc<RawHandler>,
+        });
+    }
+
+    pub(crate) fn tx_handler(
         state: &ServiceApiState,
         query: TransactionHex,
+        max_message_len: usize,
     ) -> Result<TransactionResponse, ApiError> {
-        use crate::events::error::into_failure;
-        use crate::messages::ProtocolMessage;
-
         let buf: Vec<u8> = ::hex::decode(query.tx_body).map_err(into_failure)?;
+        if buf.len() > max_message_len {
```

Comment:
Here we check exact size of message in bytes rather than payload. Because payload can contain additional 64 bytes potentially. But yes, we can just compare `query.tx_body.len() / 2` and `max_message_len` here.